frontend/app/helpers/aspace_form_helper.rb

```diff
@@ -714,29 +714,49 @@ def define_template(name, definition = nil, &block)
 
 
   def templates_for_js(jsonmodel_type = nil)
+    @delivering_js_templates = true
+
     result = ""
 
     return result if @templates.blank?
 
     obj = {}
     obj['jsonmodel_type'] = jsonmodel_type if jsonmodel_type
 
-    @templates.each do |name, template|
-      context = FormContext.new("${path}", obj, self)
+    templates_to_process = @templates.clone
+    templates_processed = []
 
-      def context.id_for(name, qualify = true)
-        name = path(name) if qualify
+    # As processing a template may register further templates that hadn't been
+    # registered previously, keep looping until we have no more templates to
+    # process.
+    while(true)
+      templates_to_process.each do |name, template|
```

Comment:
I hit an issue whereby nested templates defined new templates and these needed to be serialised to the js. So we now keep outputting js templates until there's no more to output (hence the infinite loop).